src/main/java/org/apache/accumulo/access/ParserEvaluator.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.access;
+
+import static org.apache.accumulo.access.ByteUtils.isAndOrOperator;
+
+import java.util.function.Predicate;
+
+/**
+ * Code for parsing and evaluating an access expression at the same time.
+ */
+final class ParserEvaluator {
+
+  public static final byte OPEN_PAREN = (byte) '(';
+  public static final byte CLOSE_PAREN = (byte) ')';
+  private static final byte[] EMPTY = new byte[0];
+  private static final ThreadLocal<BytesWrapper> lookupWrappers =
+      ThreadLocal.withInitial(() -> new BytesWrapper(EMPTY));
+  private static final ThreadLocal<Tokenizer> tokenizers =
+      ThreadLocal.withInitial(() -> new Tokenizer(EMPTY));
+
+  // TODO would be cleaner to have AccessEvaluatorImpl pass these in rather than use thread locals.
+  private static final Predicate<BytesWrapper> TRUE_AUTH_PREDICATE = bw -> true;
+  private static final Predicate<BytesWrapper> FALSE_AUTH_PREDICATE = bw -> false;
+
+  public static boolean parseAccessExpression(byte[] expression,
+      Predicate<BytesWrapper> authorizedPredicate) {
+
+    Tokenizer tokenizer = tokenizers.get();
+    tokenizer.reset(expression);
+
+    if (!tokenizer.hasNext()) {
+      return true;
+    }
+
+    BytesWrapper lookupWrapper = lookupWrappers.get();
+    var node = parseExpression(tokenizer, authorizedPredicate, lookupWrapper);
+
+    if (tokenizer.hasNext()) {
+      // not all input was read, so not a valid expression
+      tokenizer.error("Unexpected character '" + (char) tokenizer.peek() + "'");
+    }
+
+    return node;
+  }
+
+  private static boolean parseExpression(Tokenizer tokenizer,
+      Predicate<BytesWrapper> authorizedPredicate, BytesWrapper lookupWrapper) {
+
+    boolean result =
+        parseParenExpressionOrAuthorization(tokenizer, authorizedPredicate, lookupWrapper);
+
+    if (tokenizer.hasNext()) {
+      var operator = tokenizer.peek();
```

Comment:
`tokenizer.peek` returns a `byte`, not a `char`. When doing comparisons, use the public static members from `ByteUtils`, AND_OPERATOR and OR_OPERATOR in this case, to remove the byte to char translation.